panther_lights/src/animation/charging_animation.cpp

```diff
@@ -0,0 +1,160 @@
+  color_ = HSVtoRGB(hue / 360.0, 1.0, 1.0);
+}
+
+std::vector<uint8_t> ChargingAnimation::UpdateFrame()
+{
+  auto anim_iteration = this->GetAnimationIteration();
+  auto empty_frame = std::vector<std::uint8_t>(this->GetNumberOfLeds() * 4, 0);
+  for (std::size_t i = 3; i < empty_frame.size(); i += 4) {
+    empty_frame[i] = 255;
+  }
+
+  if (anim_iteration < fill_start_) {
+    return empty_frame;
+  }
+
+  if (anim_iteration < fill_start_ + fade_duration_) {
+    auto brightness = std::sin(M_PI_2 * (float(anim_iteration - fill_start_) / fade_duration_));
+    return CreateRGBAFrame(color_, brightness);
+  }
+
+  if (anim_iteration <= fill_end_ - fade_duration_) {
+    return CreateRGBAFrame(color_, 1.0);
+  }
+
+  if (anim_iteration < fill_end_) {
+    auto brightness = std::sin(
+      M_PI / (2.0 * fade_duration_) *
+      ((float(anim_iteration) - float(fill_end_)) + 2.0 * fade_duration_));
+    return CreateRGBAFrame(color_, brightness);
+  }
+
+  return empty_frame;
+}
+
+std::array<std::uint8_t, 3> ChargingAnimation::HSVtoRGB(
+  const float h, const float s, const float v) const
+{
+  if (std::fabs(s) < std::numeric_limits<float>::epsilon()) {
+    return {
+      {static_cast<std::uint8_t>(v * 255), static_cast<std::uint8_t>(v * 255),
+       static_cast<std::uint8_t>(v * 255)}};
+  }
+
+  const int i = static_cast<int>(std::floor(h * 6.0f));
+  const float f = (h * 6.0f) - i;
+  const float p = v * (1.0f - s);
+  const float q = v * (1.0f - s * f);
+  const float t = v * (1.0f - s * (1.0f - f));
+
+  switch (i % 6) {
+    case 0:
+      return {
+        {static_cast<std::uint8_t>(v * 255), static_cast<std::uint8_t>(t * 255),
+         static_cast<std::uint8_t>(p * 255)}};
+    case 1:
+      return {
+        {static_cast<std::uint8_t>(q * 255), static_cast<std::uint8_t>(v * 255),
+         static_cast<std::uint8_t>(p * 255)}};
+    case 2:
+      return {
+        {static_cast<std::uint8_t>(p * 255), static_cast<std::uint8_t>(v * 255),
+         static_cast<std::uint8_t>(t * 255)}};
+    case 3:
+      return {
+        {static_cast<std::uint8_t>(p * 255), static_cast<std::uint8_t>(q * 255),
+         static_cast<std::uint8_t>(v * 255)}};
+    case 4:
+      return {
+        {static_cast<std::uint8_t>(t * 255), static_cast<std::uint8_t>(p * 255),
+         static_cast<std::uint8_t>(v * 255)}};
+    default:
+      return {
+        {static_cast<std::uint8_t>(v * 255), static_cast<std::uint8_t>(p * 255),
+         static_cast<std::uint8_t>(q * 255)}};
+  }
+}
+
+std::vector<std::uint8_t> ChargingAnimation::CreateRGBAFrame(
+  const std::array<std::uint8_t, 3> color, const float brightness) const
+{
+  std::vector<std::uint8_t> frame(this->GetNumberOfLeds() * 4);
+
+  for (std::size_t i = 0; i < frame.size(); i += 4) {
+    for (int j = 0; j < 4; j++) {
+      if (j == 3) {
+        frame[i + j] = 255;
+        continue;
+      }
+      frame[i + j] = static_cast<std::uint8_t>(color[j] * brightness);
+    }
+  }
```

Comment:
```suggestion
  for (std::size_t i = 0; i < frame.size(); i += 4) {
    for (int j = 0; j < 3; j++) {
      frame[i + j] = static_cast<std::uint8_t>(color[j] * brightness);
    }
    frame[i + 3] = 255;
  }
```
Possibly something similar? or even utilize just one `for` loop?

Reply:
check the new implementation